src/query/server/query.go

```diff
@@ -830,16 +823,27 @@ func initClusters(
 			return <-dbClientCh, nil
 		}, sessionInitChan)
 
-		clustersCfg := m3.ClustersStaticConfiguration{
-			m3.ClusterStaticConfiguration{
-				Namespaces: localCfg.Namespaces,
-			},
+		clusterStaticConfig := m3.ClusterStaticConfiguration{
+			Namespaces: localCfg.Namespaces,
+		}
+		if !staticNamespaceConfig {
+			if dbCfg == nil {
+				return nil, nil, nil, errors.New("environment config required when dynamically fetching namespaces")
+			}
+			clusterStaticConfig.Client = client.Configuration{EnvironmentConfig: &dbCfg.EnvironmentConfig}
 		}
 
-		clusters, err = clustersCfg.NewStaticClusters(instrumentOpts, m3.ClustersStaticConfigurationOptions{
+		clustersCfg := m3.ClustersStaticConfiguration{clusterStaticConfig}
+		cfgOptions := m3.ClustersStaticConfigurationOptions{
 			ProvidedSession:    session,
 			CustomAdminOptions: customAdminOptions,
-		}, clusterNamespacesWatcher)
+		}
+
+		if staticNamespaceConfig {
+			clusters, err = clustersCfg.NewStaticClusters(instrumentOpts, cfgOptions, clusterNamespacesWatcher)
+		} else {
+			clusters, err = clustersCfg.NewDynamicClusters(instrumentOpts, cfgOptions, clusterNamespacesWatcher)
```

Comment:
(reading along to understand, excuse the naive-ty!) it's not clear to me why a `ClustersStaticConfiguration` would be responsible for exposing `NewDynamicClusters` if it's responsible for static config

Reply:
Heh, this is a good call out. `ClustersStaticConfiguration` should probably just be renamed `ClustersConfiguration` as the static in its name merely points out that this is a static list of clusters which is what anything is that's defined in a config file, so it's a bit redundant.

The purpose of `ClustersStaticConfiguration` is to generate a Clusters instance which is an object that can be used to get the namespaces in a cluster. A Clusters instance can be static (i.e. namespaces are defined in config) vs dynamic (i.e. namespaces are fetched from etcd). Hence the `NewStaticClusters` and `NewDynamicClusters` instance.

I should be tinkering in here over the next week or two. Will refactor and send a PR your way.